ydb/core/blobstorage/vdisk/scrub/scrub_actor_huge.cpp

```diff
@@ -27,41 +29,29 @@ namespace NKikimr {
         };
 
         auto essence = GetBarriersEssence();
-        THugeBlobMerger merger(LogPrefix, Info->Type, readHugeBlob, this);
-        TIndexRecordMerger indexMerger(Info->Type); // for GC checking
+        THugeBlobAndIndexMerger merger(LogPrefix, Info->Type, readHugeBlob, this);
 
         const TInstant startTime = TActorCoroImpl::Now();
-        do {
-            if (!iter.Valid()) {
-                break;
-            }
-
-            iter.PutToMerger(&indexMerger);
-            indexMerger.Finish();
-            auto status = essence->Keep(iter.GetCurKey(), indexMerger.GetMemRec(), {}, Snap->HullCtx->AllowKeepFlags,
-                true /*allowGarbageCollection*/);
-            indexMerger.Clear();
-
-            const TLogoBlobID& id = iter.GetCurKey().LogoBlobID();
-
-            if (status.KeepData) {
-                merger.Begin(id);
-                iter.PutToMerger(&merger);
-                const NMatrix::TVectorType needed = merger.GetPartsToRestore();
-                UpdateUnreadableParts(id, needed, merger.GetCorruptedPart());
-                if (!needed.Empty()) {
-                    Checkpoints |= TEvScrubNotify::HUGE_BLOB_SCRUBBED;
-                }
-                merger.Clear();
-            } else {
-                DropGarbageBlob(id);
-            }
-
-            iter.Prev();
-        } while (TActorCoroImpl::Now() < startTime + TDuration::Seconds(5));
-
-        if (iter.Valid()) {
-            LogoBlobIDFromLogoBlobID(iter.GetCurKey().LogoBlobID(), State->MutableBlobId());
+        if (heapIt.Valid()) {
+            auto callback = [&] (TKeyLogoBlob key, auto* merger) -> bool {
+                            auto status = essence->Keep(key, merger->GetMemRec(), {}, Snap->HullCtx->AllowKeepFlags,
+                                true /*allowGarbageCollection*/);
+                            const TLogoBlobID& id = key.LogoBlobID();
+                            if (status.KeepData) {
+                                const NMatrix::TVectorType needed = merger->GetPartsToRestore();
+                                UpdateUnreadableParts(id, needed, merger->GetCorruptedPart());
+                                if (!needed.Empty()) {
+                                    Checkpoints |= TEvScrubNotify::HUGE_BLOB_SCRUBBED;
+                                }
+                            } else {
+                                DropGarbageBlob(id);
+                            }
+                            return TActorCoroImpl::Now() < startTime + TDuration::Seconds(5);
+                        };
+            heapIt.Walk(heapIt.GetCurKey(), &merger, callback);
```

Comment:
maybe we should skip seeking here, because iterator is already positioned correctly
I suggest you adding std::optional<TKey> to Walk() and skip seeking if it is nullopt